### PR TITLE
JS: using pseudo-properties to model URL parsing

### PR DESCRIPTION
**Pull request by erik-krogh**
Now the property `searchParams` is modeled using a pseudo-property. 

Additionally I use another pseudo-property to model that calling a getter on a `URLSearchParams` retrieves the parsed parameters (see `hiddenUrlPseudoProperty` and its uses).

This means that the previous behavior is preserved, and tracking of the properties now work interprocedurally. 

E.g. with the previous local handling of `searchParams` we didn't track flow out of [this return](https://lgtm.com/projects/g/ampproject/amphtml/snapshot/894efe3a93731d65a8379c90f29862b20622f94f/files/build-system/server/routes/a4a-envelopes.js?sort=name&dir=ASC&mode=heatmap#L104) (found using [this query](https://lgtm.com/query/3328303265839020100/)).

### CODE REVIEW

File: javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll [hunk 1]
```diff
@@ -602,45 +602,89 @@ module TaintTracking {
...
     }
+
+    /**
+     * Holds if the property `prop` should be copied from the object `pred` to the object `succ`.
+     * 
+     * This step is used to copy a value the value of our pseudo-property that can later be accessed using a `get` or `getAll` call. 
+     */
+    override predicate loadStoreStep(DataFlow::Node pred, DataFlow::Node succ, string prop) {
+      prop = hiddenUrlPseudoProperty() and
+      exists(DataFlow::PropRead write | write = succ |
```
Comment (asgerf) [inline javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll]: Why is this `PropRead` in a variable called `write`? 😕

File: javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll [hunk 2]
```diff
@@ -602,45 +602,89 @@ module TaintTracking {
   }
 
   /**
-   * Holds if `params` is a `URLSearchParams` object providing access to
-   * the parameters encoded in `input`.
+   * Holds if `params` is a construction of a `URLSearchParams` that parses 
+   * the parameters in `input`.
    */
-  predicate isUrlSearchParams(DataFlow::SourceNode params, DataFlow::Node input) {
+  private predicate isUrlSearchParams(DataFlow::SourceNode params, DataFlow::Node input) {
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll]: I think I'll just remove the `private` modifier.  
The name is still fitting for the behavior, so I don't feel like adding yet another alias.

File: javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll [hunk 3]
```diff
@@ -602,45 +602,89 @@ module TaintTracking {
...
+      this = DataFlow::globalVarRef("URLSearchParams")
     }
 
+    /**
+     * Holds if `succ` is a `URLSearchParams` providing access to the 
+     * parameters encoded in `pred`. 
+     */ 
     override predicate step(DataFlow::Node pred, DataFlow::Node succ) {
-      pred = source and succ = this
+      isUrlSearchParams(succ, pred)
```
[the remark this answers is not in the document]
Comment (erik-krogh) [reply javascript/ql/src/semmle/javascript/dataflow/TaintTracking.qll]: I see two ways of solving this. 

1) Do the same as `StringManipulationTaintStep`: extend `DataFlow::ValueNode` and just bind `succ = this`. 
2) Split out the taint-steps into multiple classes, and add a characteristic predicates for each of the new classes. (Will become even more verbose). 

I went with 1) for now.

**Comment by erik-krogh** [comment]
> I found the code a bit difficult to read so for now I'll just try an relay back my understanding of the code, so we can catch any misunderstandings.
> 
> The pseudo-property represents two things lumped into a single property (which is fine):
> 
> 1. the search parameters of a URL object
> 2. the keys and values of a Map-like object (URLSearchParams)
> 
> So a property read `x.searchParams` is modelled as a load-store step of the pseudo-property in order to transition from case 1 to case 2 when accessing the search parameters of a URL object.
> 
> There's then a load step out of the pseudo-property when calling `get()`.
> 
> Assuming I got that right, I like the approach. I'd like it if you could make the code a little more accessible, though. Perhaps also add support for handle fragment data while you're at it

Your understanding is correct. 

I had the two cases lumped into the same pseudo-property because a load-store step only supported loading and storing the same property name.  
I changed it such that a load-store step can load one property and store another.  
I'm thereby better able to distinguish the two cases in the code.

I'll look into fragment data.

**Comment by erik-krogh** [comment]
An [evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-erik-krogh.northeurope.cloudapp.azure.com_1580978260567) was uneventful.

**Comment by asgerf** [comment]
Sorry for not following up on this earlier.

I like the solution, but there are a few issues that mean we're not realizing the full benefit of this change. Overall I'd like us to be able to flag this sample vuln:
```js
function getUrl() {
    return new URL(document.location);
}
$(getUrl().hash.substring(1)); // NOT OK
```

There are two issues with this at the moment:
- The `.hash` property is a [sanitizer in the Xss query](https://github.com/Semmle/ql/blob/f6af5da7f73c0533c7276a811bc408c15d571193/javascript/ql/src/semmle/javascript/security/dataflow/Xss.qll#L265), which is very restrictive, but was necessary for avoiding FPs due to variations of the `$(location.hash)` pattern. This was one of the main motivations to do more precise tracking of URLs. Due to this sanitizer, we still don't flag this vuln:

- Barriers/sanitizers currently block flow even when the tracked value is inside a property. Depending on how we resolve the above issue, this may or may not become a problem. I've discussed it with @max-schaefer and I'll experiment with changing this.

The PR would be good to land IMO, but I'd like to have some more data to verify that we're doing the right thing. I'll experiment a little bit to see what the best solution is so we can get this PR landed.

**Comment by erik-krogh** [comment]
> Overall I'd like us to be able to flag this sample vuln:
> 
> ```js
> function getUrl() {
>     return new URL(document.location);
> }
> $(getUrl().hash.substring(1)); // NOT OK
> ```

If I merge in  https://github.com/Semmle/ql/pull/2919 the above example will be flagged, and e.g. `$(window.location.hash)` is still not flagged.

But only if the source has flowlabel `"data"` (The source in the  `Xss` query has flowlabel `"taint"`). 

Here are [some examples of new flow-edges](https://lgtm.com/query/4121939901402389230/), they are not really interesting. (The results might be better once https://github.com/Semmle/ql/pull/2919 hits LGTM). 

I'll fix the merge conflict after https://github.com/Semmle/ql/pull/2919 has been merged, and I might also do another evaluation at that point.

**Comment by erik-krogh** [comment]
I did a [new evaluation](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-max.northeurope.cloudapp.azure.com_1585092548095). 
Performance looks ok, but no new results.

**Comment by erik-krogh** [comment]
> If we add this to `DomBasedXss::Configuration` we should be able to handle the `hash` example:

I'll get that in there, then I'll run one last evaluation based on that.

**Comment by erik-krogh** [comment]
> > If we add this to `DomBasedXss::Configuration` we should be able to handle the `hash` example:
> 
> I'll get that in there, then I'll run one last evaluation based on that.

[Here is an evaluation on many benchmarks just with `Xss.ql`](https://git.semmle.com/erik/dist-compare-reports/tree/profiling-js-esben.northeurope.cloudapp.azure.com_1585345841386). 
Still no new results, but performance is good.

**Pull request merged**